Speed up lost files search for large number of ignore paths

joanbm opened the pull request:
This PR implements a new option that can make the "Searching for lost files..." phase during `aconfmgr save` much faster, if a large number of ignored path patterns (`IgnorePath`) is used.

The problem with using many ignored path patterns is that those end up being passed as a large list of `-wholename` conditions to find (which in Arch Linux is GNU find), which causes the execution time of find (and hence of the entire phase) to scale with the number of ignored path patterns.

This can end up supposing the majority of the execution time. For example, I currently have 270 `IgnorePath` patterns on my configuration, and while a typical run with `--skip-checksums` but without this new option in this PR takes 35 seconds, the same run with the new option in this PR takes 15 seconds.

The way this speedup is implemented basically abuses implementation details of GNU find and has some important limitations, so honestly feel free to reject your PR if you don't want to have this junk in aconfmgr. Basically, the `-regex` option of GNU find scales much better with an increasing number of patterns (since it uses a dedicated regex library), so that's what this PR uses. The problem of course is that `-wholename` takes a shell pattern while `-regex` takes a regular expression, so we need to convert one in the other. This is pretty complex and not really a job for aconfmgr, so I've just implemented asterisk wildcards, which is sufficient for all my use cases.

The new option is called `--quick-lostfiles` and changes the lost file search behaviour in `aconfmgr save` to use the regex trick I mentioned earlier. Since it has the limitation I mentioned about only accepting asterisk wildcards, it's isn't (can't be) enabled by default.

As an additional reference I also sent a similar PR to lostfiles ( https://github.com/graysky2/lostfiles/pull/42 ) but it got stuck in a limbo.

joanbm commented:
>     * Obviously, the best place to fix it would be to improve GNU find to perform an optimization such as this implicitly. We're assuming this is not an achievable goal within the scope that we are limiting ourselves to.

Pretty much. I haven't looked at how it's implemented in GNU find, but judging from the behaviour, it looks like the way the filters are applied is just checking them in succession, with no optimization at all. While for regex they use a library which I guess does optimize common prefixes and so. So my first guess is that improving GNU find to get this case working faster would be a pretty huge task involving implementing some kind of optimizer for non-regex queries.

>     * I think the next best approach would be to detect when this optimization can be safely enabled, and then always enable it if so. If done right, no switch or other mode of configuration would be necessary.

If the idea of actually supporting all patterns doesn't pan out, this definitely sounds like a better way to go about this, rather than a new switch (not sure why it hasn't even crossed my mind).

>     * I think fully translating globs (wildcard patterns) to regular expressions is an entirely achievable task. I've put together a pure bash implementation [here](https://gist.github.com/CyberShadow/551246534146f9c2ae1b84ef6226c714) which I _believe_ to be correct.

I had in mind that shell patterns were pretty complex (if you look at `man glob`'s "GLOBBING PATTERNS" there's already a few special rules). Additionally I'm not sure what 'variant' GNU find implements since for example running `find . -wholename '*'` matches dotfiles (hidden files) while `echo *` does not.

But with some luck the rules GNU find supports (which at the end of the day are the ones we support nowadays) will be simple enough. I will take a further look to see what the actual rules are, and whether something is missing in your implementation. And as you said, fuzzing or at least testing a lot of cases is definitely the way to go to verify this.

(And thanks for putting the time for the bash implementation, that will definitely be useful as a base to work on!)
 
>     * According to `pacman -F /usr/bin/find`, GNU find is the only version of `find` which could be found on an Arch Linux system (and therefore a system that aconfmgr is running on). As such, we can assume that it will support the `-regex` switch, and that the performance is most likely going to be along the lines of what you have found.

GNU find is in findutils which is a dependency of base so I think we can safely assume it's there. I think we are already using non-standard extensions like `-wholename` so it isn't even a new problem.

>     * As far as I can see, the only new limitation that we need to be aware of with this approach is hitting the maximum argument length, which is apparently hard-coded to 128kB on Linux. Though it's an unlikely limit to hit, it may occur when ignore rules are generated automatically (usually due to the user forgetting to escape wildcards in their `IgnorePath` lines, causing them to be expanded during configuration parsing). I think it would be easy to overcome by making sure that the created regular expression is split up in arguments no longer than 131072 characters each, which could then be chained using `-o` switches like how `-wholename` rules are chained now.

I don't think this is really something someone would typically hit (and as you said it more likely to be hit due to misuse) but I'll keep this in mind.

joanbm commented:
UPDATE: It looks like GNU find just calls `fnmatch(pattern, path, 0)`, so `man fnmatch` is the right place to look, not `man glob`.

joanbm commented:
I took a look at it and found a good test suite in glibc-2.31. However, this seems to reveal that there's some more complexity that's not handled by the bash implementation, like:

* Escapes with backslash like the pattern `\<\>` should match `<>`.
* "Double brackets" like `[[:alnum:]-]` should match `-`.
* Range edge cases like `[a-c-0-9]` are valid in glob but not regex.
* i18n issues like `[a-z]` should match `ä` in `de_DE.ISO-8859-1` locale.
* Library differences like `[[.-.]]` matches `-` in glibc but not musl.

Here's the test script: https://gist.github.com/joanbm/a9277f30a2a1c8a50454ede43065544c . I don't discard some problems are due to testing with `grep -E` instead of `find` or other issues.

My general thought is that since getting the implementation right is tricky and even if doable the code will be probably be rather ugly, I'd rather go with the simpler "whitelist" approach where we check if ignore expressions are 'simple' (non-special characters, asterisk wildcards, maybe also question mark wildcards) and transform them to regex if so. An important thing to note here is that we can implement it so that it's not an "all-of-nothing" thing, so that if the user has 200 simple ignore patterns and 10 crazy ignore patterns we can 'upgrade' the 200 simple ones to `-regex` but keep the 10 crazy ones as `-wholename`.


What do you think?

CyberShadow commented:
> * Escapes with backslash like the pattern `\<\>` should match `<>`.
> 
> * "Double brackets" like `[[:alnum:]-]` should match `-`.
> 
> * Range edge cases like `[a-c-0-9]` are valid in glob but not regex.

Good finds, I missed these.

> Library differences like `[[.-.]]` matches `-` in glibc but not musl.

I think this is not an issue for our use case. I don't think Arch Linux can be made to work with another libc.

> i18n issues like `[a-z]` should match `ä` in `de_DE.ISO-8859-1` locale.

This is probably the only one that's not solvable by hammering at the bash implementation for a bit.

> An important thing to note here is that we can implement it so that it's not an "all-of-nothing" thing, so that if the user has 200 simple ignore patterns and 10 crazy ignore patterns we can 'upgrade' the 200 simple ones to `-regex` but keep the 10 crazy ones as `-wholename`.

Sounds great!

joanbm commented:
I've updated the PR with an implementation of the idea (of yours) of automatically detecting and upgrading simple patterns to regex, so the switch is no longer necessary. I'm quite satisfied of the final result, which, while still conceptually kind of a hack due to the abuse of GNU find's implementation behaviour, is self-contained and the code is relatively short and understandable.

Some notes: 

* I've chosen to only whitelist a few obviously literal characters (alphanumeric, underscore, hyphen, slash,  dot, space, hyphen, and of course the asterisk wildcards) which I expect to be the typical ones that are used in `IgnorePath`. I could have probably added more like e.g. `;=#$%&` but the benefit is minimal (they just get dropped to the 'slow' path) and I don't want to play with fire and end up with a regression.

* I played around with what you said about splitting the regular expression in chunks of size `MAX_ARG_STRLEN` (the maximum length of a single command line argument, 128K on Linux), to prevent hitting this limit prematurely if many patterns are used (and so allowing using up to `ARG_MAX`, typically 2MB on Linux). However, this won't work due to another restriction: Since we invoke `find` as root with `sudo`, and `sudo` sets the `SUDO_COMMAND` environment variable to the full command line, and enviornment variables are also limited to `MAX_ARG_STRLEN` total length, currently our effective **total** command line length limit for find is (and was) already `MAX_ARG_STRLEN`, not `ARG_MAX`. So splitting the regex by `MAX_ARG_STRLEN` is useless unless this is resolved.

CyberShadow commented on src/common.bash:
```diff
@@ -153,6 +153,51 @@ function AconfCompileOutput() {
+	# However, this is painfully slow if the number of ignore patterns is large
+	# Instead, this function (ab)uses the fact that if the number of patterns
+	# given to GNU find is big, then as an implementation detail, using regular
+	# expressions (-regex option and similar) scale smuch better than using
```
Typo:
```suggestion
	# expressions (-regex option and similar) scales much better than using
```

joanbm replied:
Will update.

CyberShadow commented on src/common.bash:
```diff
@@ -153,6 +153,51 @@ function AconfCompileOutput() {
 skip_checksums=n
 
+# Given a list of paths to ignore (which can contain shell patterns),
+# creates the list of arguments to give to 'find' to ignore them
```
```suggestion
# creates the list of arguments to give to 'find' to ignore them.
# The result is stored in the global variable "ignore_args".
```

CyberShadow replied:
There is a way to "cleanly" return an array from a bash function, without creating global variables or using pipes:

```bash
#!/bin/bash
set -eEuo pipefail

function Callee() {
	local varname=$1
	local -n var=$varname

	# shellcheck disable=SC2034
	var=(hello world)
}

function Caller() {
	local -a hello
	Callee hello
	printf '%s ' "${hello[@]}"
	printf '\n'
}

Caller

echo "$hello" # Unbound, not a global!
```

I didn't know about it when I wrote aconfmgr, which is why some places are written using less clean approaches.

joanbm replied:
Nice, I didn't know that (I not too in-depth in Bash). Will update.

CyberShadow commented on src/common.bash:
```diff
@@ -153,6 +153,51 @@ function AconfCompileOutput() {
+	local ignore_path
+	for ignore_path in "${ignore_paths[@]}"
+	do
+		if [[ "$ignore_path" =~ [^a-zA-Z0-9_/\ \-.*] ]]; then
```
Nit, the code style in aconfmgr is to use line breaks instead of semicolons (shell equivalent of Allman braces).

joanbm replied:
Will update.

CyberShadow commented on src/common.bash:
```diff
@@ -153,6 +153,51 @@ function AconfCompileOutput() {
+		# Converting the simple ignore paths to regular expressions just needs to
+		# handle the asterisk wildcard, and escape a few characters
+		local ignore_regexps
+		mapfile -t ignore_regexps < \
```
Pipe into `mapfile` instead of using process substitution. We have `lastpipe` on, so that will work and save a subshell.

joanbm replied:
Will update.

joanbm commented:
BTW, not really directly related to this PR, but I've been thinking if we could avoid all the "path rewriting" wrappers by using something like `fakeroot` or `bubblewrap` to make the tests believe they can write to the root file system (while still running as a normal user and using the binaries of the host instead of an isolated container). I'm not sure if that's really viable or if you think it makes the 'unit tests' more like 'integration tests' though.

CyberShadow commented on src/common.bash:
```diff
@@ -153,6 +153,51 @@ function AconfCompileOutput() {
+		local ignore_regexps
+		mapfile -t ignore_regexps < \
+			<( IFS=$'\n' ; echo -n "${simple_ignore_paths[*]}" | \
+			               sed 's|[^*a-zA-Z0-9_/ ]|[&]|g; s|\*|.*|g' )
```
`sed -z` + `printf '%s\0'` to allow (really) weird characters like newlines?

joanbm replied:
Doesn't matter because `simple_ignore_paths` doesn't contain any weird characters by definition (see the regex used to build it a few lines before).

CyberShadow commented on src/common.bash:
```diff
@@ -237,11 +277,11 @@ BEGIN {
 	local lost_file_count=0
 	(
 		sudo find /									\
+			 -regextype posix-extended				\
```
An explanation for why we need this particular regex type would be nice

joanbm replied:
This is mostly just to use the more modern and typical syntax, but in particular it's nice because we don't need to escape '|' alternations and '()' parenthesis. This makes code simpler since we can use `IFS` which only allows 1 character. Will add a small comment.

CyberShadow commented on test/t/lib-funcs-common.bash:
```diff
@@ -102,6 +102,9 @@ test_globals_whitelist=(
 
+	# Internal state (AconfCreateFindIgnoreArgs)
+	ignore_args
+
```
In case `local -n` doesn't work out, I think it would be better to simply unset it once it's no longer needed, as this variable has finite lifetime anyway.

joanbm replied:
`local -n` works so I will just remove this.

CyberShadow commented on test/t/lib-mocks.bash:
```diff
@@ -34,10 +34,17 @@ function find() {
+			then
+				in_regex=false
+
+				local regex_escaped_dir=$(echo -n "$test_data_dir"/files | \
+				                          sed 's/./[&]/g')
```
```suggestion
				local regex_escaped_dir=$(sed 's/./[&]/g' <<< "$test_data_dir"/files)
```

joanbm replied:
Will update.

CyberShadow commented on test/t/lib-mocks.bash:
```diff
@@ -34,10 +34,17 @@ function find() {
+
+				local regex_escaped_dir=$(echo -n "$test_data_dir"/files | \
+				                          sed 's/./[&]/g')
+				args+=("$regex_escaped_dir($arg)")
```
```suggestion
				args+=("$regex_escaped_dir($arg)") # Group |-delimited patterns
```

joanbm replied:
Will update.

CyberShadow commented:
> BTW, not really directly related to this PR, but I've been thinking if we could avoid all the "path rewriting" mocks on the unit tests by using something like `fakeroot` or `bubblewrap` to make the tests believe they can write to the root file system

Yep, a few issues here:

- `fakeroot` doesn't present a new tree to child programs, it just makes it mutable.
- `bubblewrap` can construct a new FS tree, however, the tree must contain all the userspace stuff needed for the program to run (shared libraries + loaders, CLI tools, `/dev`, ...).
- There is no running away from mocking `pacman` and friends. Even if you could get `pacman` to run in the fake root, you wouldn't be able to run it on non-Arch distros (e.g. Travis CI).
- Minimizing dependencies and maximizing execution speed are goals for the mocked test mode.
- To some extent, mocking (and thus reimplementing) the coreutils stuff helps verify our assumptions about them. There were some cases where mismatches in behavior between integration and mocked modes revealed bugs in both the test suite and aconfmgr.

joanbm commented:
PR updated with your suggestions.

Also, good to know about why unit test mocks work the way they do, this is exactly the kind of feedback I wanted to get!

CyberShadow commented:
Thank you!

The non-integration tests are failing on Travis. For some reason, it's not creating a status on the commit or PR.
https://travis-ci.org/github/CyberShadow/aconfmgr/jobs/671264120

Also, I just noticed that the test file is not named correctly. `t-1_save-2_files-4_ignored_2_shellpatterns` should be `t-1_save-2_files-4_ignored-2_shellpatterns` (dashes before numbers).

joanbm commented:
> Also, I just noticed that the test file is not named correctly.

Good catch, I've updated this for the PR.

> The non-integration tests are failing on Travis.

Well... that's very awkward, but it just helped me found a huge problem with the PR. In this line in the previous PR:

```
if [[ "$ignore_path" =~ [^a-zA-Z0-9_/\ \-.*] ]]
```

It turns out that even with a backslash, a hyphen that's not at the start or end always behaves as a range ( https://stackoverflow.com/a/55377888 ) so this regex was including more characters than expected. Due to the configuration of the Travis container this also included '?' so that's why the test failed.

Furthermore, after this I was a bit sceptical so I did some fuzz testing and generated strings with [this script](https://stackoverflow.com/a/1477572) and found a further problem which is that `a-z` included non-ASCII characters like `ä` on some locales (I guess that's obvious in retrospect). While this also worked on the regexes (at least the cases I found) it was definitely not the expected behaviour.

After I fixed the regex by enumerating each ASCII character instead of using a range, I could find no more problems by fuzzing and all tests now pass on my Travis container, so I've updated the PR with this.

CyberShadow commented:
> For some reason, it's not creating a status on the commit or PR.

I unauthorized the Travis app from my GitHub account and logged in on its website, and it seems to be creating commit statuses again.

joanbm commented:
> Thank you very much again for this improvement and the thoroughness of your work.

Thank you so much as well! This took more time than I expected to get right, so I appreciate all the time you've also spent to make this happen.
